src/html5_parser/parser/adoption_agency.rs

```diff
@@ -195,80 +188,83 @@ impl<'a> Html5Parser<'a> {
-
-                if node.parent.is_some() {
-                    node.parent = None
-                }
+                let afe_idx = self
+                    .active_formatting_elements
+                    .iter()
+                    .position(|elem| elem == &ActiveElement::Node(node_id))
+                    .expect("node not found");
+                self.active_formatting_elements[afe_idx] = ActiveElement::Node(replacement_node_id);
+                let idx = self
+                    .open_elements
+                    .iter()
+                    .position(|elem| elem == &node_id)
+                    .expect("node not found");
+                self.open_elements[idx] = replacement_node_id;
 
                 let node_id = replacement_node_id;
-                // let node = self.document.get_node_by_id(node_id).expect("node not found").clone();
 
                 // Step 4.13.7
-                if last_node_idx == furthest_block_idx {
-                    bookmark = node_idx - 1;
+                if last_node_id == furthest_block_id {
+                    bookmark_afe = afe_idx + 1;
                 }
 
                 // Step 4.13.8
-                self.document.append(last_node_id, node_id);
+                self.document.relocate(last_node_id, node_id);
 
                 // Step 4.13.9
                 last_node_id = node_id;
             }
 
             // Step 4.14
-            self.document.append(last_node_id, common_ancestor_id);
+            self.document.relocate(last_node_id, common_ancestor_id);
 
             // Step 4.15
             let new_element = match formatting_element_node.data {
-                NodeData::Element { ref attributes, .. } => Node::new_element(
-                    formatting_element_node.name.as_str(),
-                    attributes.clone(),
-                    HTML_NAMESPACE,
-                ),
+                NodeData::Element {
+                    ref name,
+                    ref attributes,
+                    ..
+                } => Node::new_element(name.as_str(), attributes.clone(), HTML_NAMESPACE),
                 _ => panic!("formatting element is not an element"),
             };
 
+            // Step 4.17
+            let new_element_id = self.document.add_node(new_element, furthest_block_id);
+
             // Step 4.16
-            for &child in furthest_block_node.children.iter() {
-                self.document.append(child, new_element.id)
+            for child in furthest_block_node.children.iter() {
+                self.document.relocate(*child, new_element_id);
             }
-            furthest_block_node.children.clear();
-
-            // Step 4.17
-            let new_element_id = self.document.add_node(new_element, furthest_block_node.id);
 
             // Step 4.18
             self.active_formatting_elements
-                .remove(formatting_element_idx);
+                .insert(bookmark_afe, ActiveElement::Node(new_element_id));
```

Comment:
just curious, what's the difference when swapping the order of the insert/remove here?

Reply:
when adding the element on the bookmark, the `formatting_element_idx_afe` index (the index of the formatting element in the active formatting elements stack (afe), might have changed.

But there still might be an issue with this, so this might still be missing a check when the formatting_element_idx_afe > bookmark_idx_afe..    This would explain some issues I still see

Reply:
I've fixed that issue as well in the latest version